### PR TITLE
Support duplicate prefixes in long column names for CSV uploads

### PR DESCRIPTION
**Pull request by crisptrutski**
### Description

Work backwards from https://github.com/metabase/metabase/pull/44809 to see the discussion that led here. This restores support for initial uploads with duplicate long columns, after it was broken [here](https://github.com/metabase/metabase/pull/44873), and opens the door to supporting appends as well.

### CODE REVIEW

**Comment by crisptrutski** [comment]
It seems like this has actually broken `models-metadata.cy.spec.js` (failed 3 times)

File: src/metabase/upload.clj
Comment (calherries) [inline src/metabase/upload.clj:256]: wdyt about pulling this out of `lib.util`? The fact it's in a `metabase.lib.util` namespace might have the QP team thinking they can just modify it without breaking CSV upload behaviour. I've started doing that in https://github.com/metabase/metabase/pull/44862. To minimize conflicts maybe it can wait until both these PRs have landed though

**Review by calherries — APPROVED** [review]
Great, this is nice. Kudos to your persistence on solving this problem, I like it :D

File: test/metabase/upload_test.clj
Comment (calherries) [inline test/metabase/upload_test.clj:2074]: Some test coverage with unicode characters wouldn't be so hard to add, and seems the most likely thing to forget about and break in the future

File: src/metabase/upload.clj
Comment (crisptrutski) [reply src/metabase/upload.clj:256]: Yeah I like that direction a lot. The way this uses legacy and current lib functions together also seems odd - perhaps we can find a shared home for them.

Definitely let's wait for dust to settle a bit before we do the move.

File: test/metabase/upload_test.clj
Comment (crisptrutski) [reply test/metabase/upload_test.clj:2074]: I've made a note to update these tests, and the pre-existing ones, in a later PR on Monday. Will go nicely on top of the display stuff you worked on.

Comment (crisptrutski) [reply test/metabase/upload_test.clj:2074]: Oh well, there are merge conflicts now anyway, if CI needs to run again I'll just update this before merging.

**Pull request merged**
